odl/solvers/functional/functional.py

```diff
@@ -0,0 +1,1471 @@
+            raise TypeError('functional {!r} is not a Functional instance.'
+                            ''.format(func))
+
+        scalar = func.range.element(scalar)
+
+        if scalar > 0:
+            Functional.__init__(self, domain=func.domain,
+                                linear=func.is_linear,
+                                smooth=func.is_smooth,
+                                concave=func.is_concave,
+                                convex=func.is_convex,
+                                grad_lipschitz=(scalar * func.grad_lipschitz))
+        elif scalar == 0:
+            Functional.__init__(self, domain=func.domain,
+                                linear=True,
+                                smooth=True,
+                                concave=True, convex=True,
+                                grad_lipschitz=0)
+        elif scalar < 0:
+            Functional.__init__(self, domain=func.domain,
+                                linear=func.is_linear,
+                                smooth=func.is_smooth,
+                                concave=func.is_convex,
+                                convex=func.is_concave,
+                                grad_lipschitz=(-scalar * func.grad_lipschitz))
+        else:
+            # It should not be possible to get here
+            raise TypeError('comparison with scalar {} failed'.format(scalar))
+
+        OperatorLeftScalarMult.__init__(self, operator=func, scalar=scalar)
+
+        self._func = func
+        self._scalar = scalar
+
+    @property
+    def gradient(self):
+        """Gradient operator of the functional.
+
+        Notes
+        -----
+        The operator that corresponds to the mapping
+
+        .. math::
+
+            x \\to \\nabla f(x)
+
+        where :math:`\\nabla f(x)` is the element used to evaluated
+        derivatives in a direction :math:`d` by
+        :math:`\\langle \\nabla f(x), d \\rangle`.
+        """
+
+        functional = self
+
+        class LeftScalarMultGradient(Operator):
+            """The gradient operator for a `FunctionalLeftScalarMult`
+            functional.
+            """
+            def __init__(self):
+                """Initialize a `LeftScalarMultGradient` instance."""
+                super().__init__(functional.domain, functional.domain,
+                                 linear=False)
+
+            def _call(self, x):
+                """Evaluates the gradient in a point ``x``.
+
+                Parameters
+                ----------
+                x : `LinearSpaceVector`
+                    Element in the domain of the operator.
+
+                Returns
+                -------
+                out : `LinearSpaceVector`
+                    The gradient in the point ``x``. An element in the doimain
+                    of the operator.
+                """
+                return functional._scalar * functional._func.gradient(x)
+
+        return LeftScalarMultGradient()
```

Comment:
Why not simply remove the class and do

`return self.scalar * self.operator.gradient`